openrtb_ext/request.go

```diff
@@ -314,3 +317,178 @@ func (m ExtMultiBid) String() string {
 	}
 	return fmt.Sprintf("{Bidder:%s, Bidders:%v, MaxBids:%s, TargetBidderCodePrefix:%s}", m.Bidder, m.Bidders, maxBid, m.TargetBidderCodePrefix)
 }
+
+func (erp *ExtRequestPrebid) Clone() *ExtRequestPrebid {
+	if erp == nil {
+		return nil
+	}
+
+	clone := *erp
+	clone.Aliases = maputil.CloneMap[string, string](erp.Aliases)
+	clone.AliasGVLIDs = maputil.CloneMap[string, uint16](erp.AliasGVLIDs)
+	clone.BidAdjustmentFactors = maputil.CloneMap[string, float64](erp.BidAdjustmentFactors)
+
+	if erp.BidderConfigs != nil {
+		newBidderConfigs := make([]BidderConfig, len(erp.BidderConfigs))
+		for i, bc := range erp.BidderConfigs {
+			newBidderConfig := BidderConfig{Bidders: sliceutil.CloneSlice[string](bc.Bidders)}
+			if bc.Config != nil {
+				config := &Config{}
+				if bc.Config.ORTB2 != nil {
+					config.ORTB2 = &ORTB2{
+						Site: maputil.CloneMap(bc.Config.ORTB2.Site),
+						App:  maputil.CloneMap(bc.Config.ORTB2.App),
+						User: maputil.CloneMap(bc.Config.ORTB2.User),
+					}
+				}
+				newBidderConfig.Config = config
+			}
+			newBidderConfigs[i] = newBidderConfig
+		}
+		clone.BidderConfigs = newBidderConfigs
+	}
+
+	if erp.Cache != nil {
+		clone.Cache = &ExtRequestPrebidCache{}
+		if erp.Cache.Bids != nil {
+			clone.Cache.Bids = &ExtRequestPrebidCacheBids{}
+			if erp.Cache.Bids.ReturnCreative != nil {
+				clone.Cache.Bids.ReturnCreative = ptrutil.ToPtr[bool](*erp.Cache.Bids.ReturnCreative)
+			}
+		}
+		if erp.Cache.VastXML != nil {
+			clone.Cache.VastXML = &ExtRequestPrebidCacheVAST{}
+			if erp.Cache.VastXML.ReturnCreative != nil {
+				clone.Cache.VastXML.ReturnCreative = ptrutil.ToPtr[bool](*erp.Cache.VastXML.ReturnCreative)
+			}
+		}
+	}
+
+	if erp.Channel != nil {
+		channel := *erp.Channel
+		clone.Channel = &channel
+	}
+
+	if erp.CurrencyConversions != nil {
+		newConvRates := make(map[string]map[string]float64, len(erp.CurrencyConversions.ConversionRates))
+		for key, val := range erp.CurrencyConversions.ConversionRates {
+			newConvRates[key] = maputil.CloneMap(val)
+		}
+		clone.CurrencyConversions = &ExtRequestCurrency{ConversionRates: newConvRates}
+		if erp.CurrencyConversions.UsePBSRates != nil {
+			clone.CurrencyConversions.UsePBSRates = ptrutil.ToPtr[bool](*erp.CurrencyConversions.UsePBSRates)
+		}
+	}
+
+	if erp.Data != nil {
+		clone.Data = &ExtRequestPrebidData{Bidders: sliceutil.CloneSlice(erp.Data.Bidders)}
+		if erp.Data.EidPermissions != nil {
+			newEidPermissions := make([]ExtRequestPrebidDataEidPermission, 0, len(erp.Data.EidPermissions))
+			for _, eidp := range erp.Data.EidPermissions {
```

Comment:
Assuming your suggestion is coming from a performance perspective, I think the usual performance hit due to allocations when using `append` won't apply here because we have already specified the length and capacity of the slice on L386 using the `make` function. So performance wise this should be similar to using indexes. However, I do think it is useful here to use indexes rather than `append` for consistency sake.

Reply:
> Assuming your suggestion is coming from a performance perspective,

It was not. My question was for consistency.

Reply:
Yes, this could use some cleaning up.